### ISSUE TITLE
OP_ENTERSUB is not called for constructor of tied object

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16382]
Migrated from [rt.perl.org#132768](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132768) (status was 'rejected')

Searchable as RT132768$

**Comment by p5pRT** [comment]
### From @KES777

#### Created by @KES777



TLDR;
When we tie scalar to 'MyObj' class the MyObj&#8203;::TIESCALAR is called
but OP\_ENTERSUB is not executed before that\. 
But still we can see OP\_LEAVESUB for it


How to reproduce&#8203;:
$cat script\.pl
package Obj;
sub DESTROY \{
&nbsp;    print "DESTROY\\n";
\}

sub new\{ bless \{\} \};
package MyTie;

our $test;

sub TIESCALAR \{
&nbsp;    bless \{\}\, 'MyTie&#8203;::Scalar';
\}

package MyTie&#8203;::Scalar;

package main;

sub MyTie&#8203;::Scalar&#8203;::FETCH \{
&nbsp;    local $MyTie&#8203;::test = Obj\->new;
&nbsp;    print "FETCHed\\n";
&nbsp;    return 7;
\}


my $x;
tie $x\, 'MyTie';
\{
&nbsp;    local $MyTie&#8203;::test = 3;
&nbsp;    print $x;

&nbsp;    print "TEST&#8203;: $MyTie&#8203;::test\\n";
\}


$cat script\.xs
\#define PERL\_NO\_GET\_CONTEXT
\#include "EXTERN\.h"
\#include "perl\.h"
\#include "XSUB\.h"

\#include "ppport\.h"

int
My\_runops2\(pTHX\)\{
&nbsp;    SV\* id =  get\_sv\("DB&#8203;::runops\_id"\, GV\_ADD|GV\_ADDMULTI\);
&nbsp;    save\_item\( id \);
&nbsp;    sv\_inc\_nomg\( id \);
&nbsp;    int idv = SvIV\( id \);

&nbsp;    printf\( "Run OP start\(%d\)\\n"\, idv \);

&nbsp;    SV \*tv =  newSVpvs\( "MyTie&#8203;::Scalar" \);

&nbsp;    OP \*o; CV \*cv; SV \*pkg;
&nbsp;    while\( PL\_op \) \{
&nbsp;        o =  PL\_op;

&nbsp;        SV\* intv =  get\_sv\("DB&#8203;::intv"\, 0\);
&nbsp;        printf\( "OP Type&#8203;: %d \-\- %s intv&#8203;:%d\\n"\, o\->op\_type\, OP\_DESC\( o \)\, SvTRUE\( intv \) \);

&nbsp;        switch\( o\->op\_type \) \{
&nbsp;        case OP\_ENTERSUB&#8203;:;
&nbsp;            PL\_op = o =  CALL\_FPTR\(o\->op\_ppaddr\)\(aTHX\);

&nbsp;            cv =  find\_runcv\(0\);
&nbsp;            pkg = newSVpv\( HvNAME\( GvSTASH\(CvGV\(cv\)\) \)\, 0 \);
&nbsp;            printf\( "ENTER&#8203;: %s&#8203;::%s\\n"\, SvPV\_nolen\( pkg \)\, GvNAME\( CvGV\(cv\) \) \);

&nbsp;            // sv\_dump\( cv \);
&nbsp;            break;
&nbsp;        case OP\_LEAVESUB&#8203;:;

&nbsp;            cv =  find\_runcv\(0\);
&nbsp;            pkg = newSVpv\( HvNAME\( GvSTASH\(CvGV\(cv\)\) \)\, 0 \);
&nbsp;            printf\( "LEAVE&#8203;: %s&#8203;::%s\\n"\, SvPV\_nolen\( pkg \)\, GvNAME\( CvGV\(cv\) \) \);


&nbsp;            // sv\_dump\( cv \);
&nbsp;            break;
&nbsp;        \}


&nbsp;        PL\_op = CALL\_FPTR\(o\->op\_ppaddr\)\(aTHX\);
&nbsp;    \}

&nbsp;    printf\( "Run OP finish \(%d\)\\n"\, idv \);

&nbsp;    PERL\_ASYNC\_CHECK\(\);
&nbsp;    TAINT\_NOT;
&nbsp;    return 0;
\}

MODULE = Module                    PACKAGE = Module
BOOT&#8203;:
&nbsp;    get\_sv\("DB&#8203;::intv"\,   GV\_ADD|GV\_ADDMULTI\);

&nbsp;    PL\_runops =  My\_runops2;

$perl \-MModule script\.pl
Run OP start\(1\)
OP Type&#8203;: 0 \-\- null operation intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 68 \-\- string intv&#8203;:0
Run OP finish \(1\)
Run OP start\(1\)
OP Type&#8203;: 0 \-\- null operation intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 68 \-\- string intv&#8203;:0
Run OP finish \(1\)
Run OP start\(1\)
OP Type&#8203;: 196 \-\- block entry intv&#8203;:0
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 9 \-\- private variable intv&#8203;:0
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 9 \-\- private variable intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 226 \-\- tie intv&#8203;:0

/\*\*\* here TIESCALAR is called but there is no OP\_ENTERSUB for it \*\*\*/
Run OP start\(2\)
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 161 \-\- anonymous hash \(\{\}\) intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 23 \-\- bless intv&#8203;:0
OP Type&#8203;: 186 \-\- subroutine exit intv&#8203;:0
LEAVE&#8203;: MyTie&#8203;::TIESCALAR
Run OP finish \(2\)

/\*\*\* Actually next is not relevant\, but provided for completeness \*\*\*/

OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 201 \-\- loop entry intv&#8203;:0
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 6 \-\- scalar variable intv&#8203;:0
OP Type&#8203;: 37 \-\- scalar assignment intv&#8203;:0
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 9 \-\- private variable intv&#8203;:0
OP Type&#8203;: 238 \-\- print intv&#8203;:0
Run OP start\(2\)
OP Type&#8203;: 210 \-\- method with known name intv&#8203;:0
OP Type&#8203;: 185 \-\- subroutine entry intv&#8203;:0
ENTER&#8203;: MyTie&#8203;::Scalar&#8203;::FETCH
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 210 \-\- method with known name intv&#8203;:0
OP Type&#8203;: 185 \-\- subroutine entry intv&#8203;:0
ENTER&#8203;: Obj&#8203;::new
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 161 \-\- anonymous hash \(\{\}\) intv&#8203;:0
OP Type&#8203;: 23 \-\- bless intv&#8203;:0
OP Type&#8203;: 186 \-\- subroutine exit intv&#8203;:0
LEAVE&#8203;: Obj&#8203;::new
OP Type&#8203;: 6 \-\- scalar variable intv&#8203;:0
OP Type&#8203;: 37 \-\- scalar assignment intv&#8203;:0
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 238 \-\- print intv&#8203;:0
FETCHed
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 186 \-\- subroutine exit intv&#8203;:0
LEAVE&#8203;: MyTie&#8203;::Scalar&#8203;::FETCH
Run OP start\(3\)
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 238 \-\- print intv&#8203;:0
DESTROY
OP Type&#8203;: 186 \-\- subroutine exit intv&#8203;:0
LEAVE&#8203;: Obj&#8203;::DESTROY
Run OP finish \(3\)
Run OP finish \(2\)
OP Type&#8203;: 193 \-\- next statement intv&#8203;:0
OP Type&#8203;: 3 \-\- pushmark intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 6 \-\- scalar variable intv&#8203;:0
OP Type&#8203;: 67 \-\- concatenation \(\.\) or string intv&#8203;:0
OP Type&#8203;: 5 \-\- constant item intv&#8203;:0
OP Type&#8203;: 67 \-\- concatenation \(\.\) or string intv&#8203;:0
OP Type&#8203;: 238 \-\- print intv&#8203;:0
7TEST&#8203;: 3
OP Type&#8203;: 202 \-\- loop exit intv&#8203;:0
OP Type&#8203;: 197 \-\- block exit intv&#8203;:0
Run OP finish \(1\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.24.0:

Configured by kes at Wed Oct 19 14:07:47 EEST 2016.

Summary of my perl5 (revision 5 version 24 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=4.4.0-43-generic, archname=x86_64-linux
    uname='linux work 4.4.0-43-generic #63-ubuntu smp wed oct 12 13:48:03 utc 2016 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/kes/perl5/perlbrew/perls/perl-5.24.0 -Aeval:scriptdir=/home/kes/perl5/perlbrew/perls/perl-5.24.0/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='5.4.0 20160609', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /lib64 /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.23.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.23'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'

Locally applied patches:
    Devel::PatchPerl 1.38


@INC for perl 5.24.0:
    /home/kes/perl5/perlbrew/perls/perl-5.24.0/lib/site_perl/5.24.0/x86_64-linux
    /home/kes/perl5/perlbrew/perls/perl-5.24.0/lib/site_perl/5.24.0
    /home/kes/perl5/perlbrew/perls/perl-5.24.0/lib/5.24.0/x86_64-linux
    /home/kes/perl5/perlbrew/perls/perl-5.24.0/lib/5.24.0
    .


Environment for perl 5.24.0:
    HOME=/home/kes
    LANG=en_US.UTF-8
    LANGUAGE=en
    LC_ADDRESS=uk_UA.UTF-8
    LC_IDENTIFICATION=uk_UA.UTF-8
    LC_MEASUREMENT=uk_UA.UTF-8
    LC_MESSAGES=en_US.UTF-8
    LC_MONETARY=uk_UA.UTF-8
    LC_NAME=uk_UA.UTF-8
    LC_NUMERIC=uk_UA.UTF-8
    LC_PAPER=uk_UA.UTF-8
    LC_TELEPHONE=uk_UA.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/kes/perl5/perlbrew/bin:/home/kes/perl5/perlbrew/perls/perl-5.24.0/bin:/home/kes/bin:/home/kes/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERLBREW=command perlbrew
    PERLBREW_BASHRC_VERSION=0.78
    PERLBREW_HOME=/home/kes/.perlbrew
    PERLBREW_MANPATH=/home/kes/perl5/perlbrew/perls/perl-5.24.0/man
    PERLBREW_PATH=/home/kes/perl5/perlbrew/bin:/home/kes/perl5/perlbrew/perls/perl-5.24.0/bin
    PERLBREW_PERL=perl-5.24.0
    PERLBREW_ROOT=/home/kes/perl5/perlbrew
    PERLBREW_VERSION=0.78
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jan 25\, 2018 at 10&#8203;:00&#8203;:29AM \-0800\, KES wrote&#8203;:
> When we tie scalar to 'MyObj' class the MyObj&#8203;::TIESCALAR is called
> but OP\_ENTERSUB is not executed before that\. 
> But still we can see OP\_LEAVESUB for it

pp\_entersub\(\) is still called\, but it is called directly\, rather than from
a runops loop\. Is this a problem for you?

\-\- 
In economics\, the exam questions are the same every year\.
They just change the answers\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @KES777

yes\, it is\. Because when I count depth level it occasionally became negative\.
So it is better to see both corresponding events\.

Also I have noticed that for the DESTROY method pp\_entersub is not called from a runops loop too\.

**Comment by p5pRT** [comment]
### From zefram@fysh.org

KES wrote&#8203;:
>When we tie scalar to 'MyObj' class the MyObj&#8203;::TIESCALAR is called
>but OP\_ENTERSUB is not executed before that\. 

Not a bug\.  There is no entersub op because the sub call is not being
made from Perl code\.  There are many places in which subs get called by
means other than explicit calls from Perl code\.  Any XS code can make
a sub call\, and we even have a whole documentation file advising XS
programmers how to do that\.  It is not remotely viable to rely on sub
calls being made only through Perl ops\.

>But still we can see OP\_LEAVESUB for it

The leavesub op isn't part of the call site\, it's part of the called sub\.
Of course it only exists for a called Perl sub\.  You'll see the opposite
pattern when calling an XS sub explicitly from Perl code&#8203;: an entersub
op at the call site\, then no leavesub or any other op until control has
returned to the call site and moved on to the next thing\.

\-zefram

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Jan 28\, 2018 at 01&#8203;:03&#8203;:57PM \+0000\, Zefram wrote&#8203;:
> KES wrote&#8203;:
> >When we tie scalar to 'MyObj' class the MyObj&#8203;::TIESCALAR is called
> >but OP\_ENTERSUB is not executed before that\. 
> 
> Not a bug\.  There is no entersub op because the sub call is not being
> made from Perl code\.  There are many places in which subs get called by
> means other than explicit calls from Perl code\.  Any XS code can make
> a sub call\, and we even have a whole documentation file advising XS
> programmers how to do that\.  It is not remotely viable to rely on sub
> calls being made only through Perl ops\.
> 
> >But still we can see OP\_LEAVESUB for it
> 
> The leavesub op isn't part of the call site\, it's part of the called sub\.
> Of course it only exists for a called Perl sub\.  You'll see the opposite
> pattern when calling an XS sub explicitly from Perl code&#8203;: an entersub
> op at the call site\, then no leavesub or any other op until control has
> returned to the call site and moved on to the next thing\.

I agree that this ticket should be rejected\.

\-\- 
"I do not resent criticism\, even when\, for the sake of emphasis\,
it parts for the time with reality"\.
&nbsp;    \-\- Winston Churchill\, House of Commons\, 22nd Jan 1941\.

**Comment by p5pRT** [comment]
@xsawyerx - Status changed from 'open' to 'rejected'
